bpo-40394 - difflib.SequenceMatched.find_longest_match default args

lrjball opened the pull request:
Added default args to find_longest_match, as well as related tests

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-40394](https://bugs.python.org/issue40394) -->
https://bugs.python.org/issue40394
<!-- /issue-number -->

lrjball committed:
bpo-40394 - difflib.SequenceMatched.find_longest_match default args

Added default args to find_longest_match, as well as related tests

tim-one commented on Lib/test/test_difflib.py:
```diff
@@ -501,12 +501,30 @@ def test_is_character_junk_false(self):
+        match = sm.find_longest_match()
+        self.assertEqual(match.a, 0)
+        self.assertEqual(match.b, 0)
+        self.assertEqual(match.size, 6)
```
Would be good to also test a mix of explicit and default args.  For example, like this test, but with `alo=2` and `bhi=6`.

tim-one reviewed (changes requested):
Looks good so far!  Needs more stuff:

- Add your name to `Misc/ACKS`
- Docs need to be updated to match, and with a `versionchanged` note.
- Needs a `NEWS` entry.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

lrjball committed:
Added docs, ACKS and more thorough tests

- Added docs to difflib.rst
- Added entry to ACKS
- Added more thorough testing by checking various combinations of the
args, and checking that it does give the actual longest match

lrjball committed:
Merge branch 'fix-issue-40394' of https://github.com/lrjball/cpython into fix-issue-40394

tim-one commented on Lib/difflib.py:
```diff
@@ -391,6 +393,8 @@ def find_longest_match(self, alo, ahi, blo, bhi):
 
         a, b, b2j, isbjunk = self.a, self.b, self.b2j, self.bjunk.__contains__
+        ahi = len(a) if ahi is None else ahi
+        bhi = len(b) if bhi is None else bhi
```
```suggestion
        if ahi is None:
            ahi = len(a)
        if bhi is None:
            bhi = len(b)
```
It's a small point, but this is easier to follow the suggested way - which is the way similar defaults are resolved throughout the code base.  The "extra" lines don't matter at all.  What does matter is that it's visually obvious at once that "`is None`" is the _only_ thing going on.

tim-one reviewed (changes requested):
Nice!  I made one more code suggestion, which you should be able to commit directly from the comment.  Or argue about it, but I don't suggest that :wink:

lrjball committed:
Update Lib/difflib.py

Co-Authored-By: Tim Peters <tim.peters@gmail.com>

tim-one reviewed (approved):
All good - thank you, Lewis!  I'll merge this now :smiley:
